pkg/fetcher/fetcher.go

```diff
@@ -280,13 +296,32 @@ func (u *URLFetcher) fetch(ctx context.Context, url *url.URL, reqHdrs *http.Head
 			}
 		}
 
-		err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s.", http.StatusText(u.StatusCode), u.StatusCode, url)
+		// for all other non-retryable client error, grab the error message if there is one (#5951)
+		msg, err := u.extractErrResponseMessage(res.Body)
+		if err != nil {
+			log.Debugf("Nonretryable client error in unconventional format: %s", err)
+			err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s", http.StatusText(u.StatusCode), u.StatusCode, url)
+		} else {
+			err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s, Message: %s", http.StatusText(u.StatusCode),
+				u.StatusCode, url, msg)
+		}
+
 		return nil, nil, DoNotRetry{Err: err}
 	}
 
 	// FIXME: handle StatusTemporaryRedirect and StatusFound
+	// for all other unexpected http code, grab the message out if there is one (#5951)
 	if !u.IsStatusOK() {
-		return nil, nil, fmt.Errorf("Unexpected http code: %d, URL: %s", u.StatusCode, url)
+
+		msg, err := u.extractErrResponseMessage(res.Body)
```

Comment:
yes but before i worked on this issue, we're differentiating nonretryable error (4xx http code) from all other errors (5xx or sth else), and we're outputing messages differently, so that's why i did the same thing...
if we don't want the user to get the verbose `nonretryable error` message, i can just squash this condition and put it together with everything else